botocore/useragent.py

```diff
@@ -94,11 +141,40 @@ def to_string(self):
             self.name, allow_hash=False
         )
         if self.value is None or self.value == '':
-            return f'{clean_prefix}/{clean_name}'
-        clean_value = sanitize_user_agent_string_component(
-            self.value, allow_hash=True
-        )
-        return f'{clean_prefix}/{clean_name}#{clean_value}'
+            clean_string = f'{clean_prefix}/{clean_name}'
+        else:
+            clean_value = sanitize_user_agent_string_component(
+                self.value, allow_hash=True
+            )
+            clean_string = f'{clean_prefix}/{clean_name}#{clean_value}'
+        if self.size_config is not None:
+            clean_string = self._truncate_string(
+                clean_string,
+                self.size_config.max_size_in_bytes,
+                self.size_config.delimiter,
+            )
+        return clean_string
+
+    def _truncate_string(self, string, max_size, delimiter):
+        """
+        Pop ``delimiter``-separated values until encoded string is less than or
+        equal to ``max_size``.
+        """
+        orig = string
+        encoded_string = string.encode('utf-8')
+        while len(encoded_string) > max_size:
```

Comment:
nit - Passing in a negative value for `max_size` would result in an infinite loop. Kind of a silly thing to guard against since we control what gets configured provided. I don't feel strongly about it, but maybe something to consider.

Reply:
Addressed in https://github.com/boto/botocore/pull/3389/commits/a743c33783692038adf7f573449eab721e06fab2